Android Proguard - Can't find references method 'kategory.EitherT fromEither

Hi, 

First of all thanks for the library, I'm trying to get the hand of it :)

For now, When I'm trying to compile with proguard I get this error.

```
Warning:kategory.EitherT: can't find referenced method 'kategory.EitherT fromEither(kategory.Either,kategory.Applicative)' in program class kategory.EitherT$Companion
Warning:kategory.Kleisli: can't find referenced method 'kategory.Kleisli pure(java.lang.Object,kategory.Applicative)' in program class kategory.Kleisli$Companion
Warning:kategory.Kleisli: can't find referenced method 'kategory.Kleisli ask(kategory.Applicative)' in program class kategory.Kleisli$Companion
Warning:kategory.OptionT: can't find referenced method 'kategory.OptionT pure(java.lang.Object,kategory.Applicative)' in program class kategory.OptionT$Companion
Warning:kategory.OptionT: can't find referenced method 'kategory.OptionT none(kategory.Applicative)' in program class kategory.OptionT$Companion
Warning:kategory.OptionT: can't find referenced method 'kategory.OptionT fromOption(kategory.Option,kategory.Applicative)' in program class kategory.OptionT$Companion
Warning:there were 6 unresolved references to program class members.
Warning:Exception while processing task java.io.IOException: Please correct the above warnings first.
Error:Execution failed for task ':app:transformClassesAndResourcesWithProguardForProdRelease'.
> Job failed, see logs for details
```

I have only used `implementation 'io.kategory:kategory:0.3.11'` in my app.gradle with Android Studio 3.0

Any idea how I can solve this?

```
* Exception is:
org.gradle.api.tasks.TaskExecutionException: Execution failed for task ':app:transformClassesAndResourcesWithProguardForDevRelease'.
	at org.gradle.api.internal.tasks.execution.ExecuteActionsTaskExecuter.executeActions(ExecuteActionsTaskExecuter.java:100)
	at org.gradle.api.internal.tasks.execution.ExecuteActionsTaskExecuter.execute(ExecuteActionsTaskExecuter.java:70)
	at org.gradle.api.internal.tasks.execution.SkipCachedTaskExecuter.execute(SkipCachedTaskExecuter.java:106)
	at org.gradle.api.internal.tasks.execution.SkipUpToDateTaskExecuter.execute(SkipUpToDateTaskExecuter.java:63)
	at org.gradle.api.internal.tasks.execution.ResolveTaskOutputCachingStateExecuter.execute(ResolveTaskOutputCachingStateExecuter.java:54)
	at org.gradle.api.internal.tasks.execution.ResolveBuildCacheKeyExecuter.execute(ResolveBuildCacheKeyExecuter.java:60)
	at org.gradle.api.internal.tasks.execution.ValidatingTaskExecuter.execute(ValidatingTaskExecuter.java:58)
	at org.gradle.api.internal.tasks.execution.SkipEmptySourceFilesTaskExecuter.execute(SkipEmptySourceFilesTaskExecuter.java:88)
	at org.gradle.api.internal.tasks.execution.ResolveTaskArtifactStateTaskExecuter.execute(ResolveTaskArtifactStateTaskExecuter.java:52)
	at org.gradle.api.internal.tasks.execution.SkipTaskWithNoActionsExecuter.execute(SkipTaskWithNoActionsExecuter.java:52)
	at org.gradle.api.internal.tasks.execution.SkipOnlyIfTaskExecuter.execute(SkipOnlyIfTaskExecuter.java:54)
	at org.gradle.api.internal.tasks.execution.ExecuteAtMostOnceTaskExecuter.execute(ExecuteAtMostOnceTaskExecuter.java:43)
	at org.gradle.api.internal.tasks.execution.CatchExceptionTaskExecuter.execute(CatchExceptionTaskExecuter.java:34)
	at org.gradle.execution.taskgraph.DefaultTaskGraphExecuter$EventFiringTaskWorker$1.run(DefaultTaskGraphExecuter.java:248)
	at org.gradle.internal.progress.DefaultBuildOperationExecutor$RunnableBuildOperationWorker.execute(DefaultBuildOperationExecutor.java:336)
	at org.gradle.internal.progress.DefaultBuildOperationExecutor$RunnableBuildOperationWorker.execute(DefaultBuildOperationExecutor.java:328)
	at org.gradle.internal.progress.DefaultBuildOperationExecutor.execute(DefaultBuildOperationExecutor.java:197)
	at org.gradle.internal.progress.DefaultBuildOperationExecutor.run(DefaultBuildOperationExecutor.java:107)
	at org.gradle.execution.taskgraph.DefaultTaskGraphExecuter$EventFiringTaskWorker.execute(DefaultTaskGraphExecuter.java:241)
	at org.gradle.execution.taskgraph.DefaultTaskGraphExecuter$EventFiringTaskWorker.execute(DefaultTaskGraphExecuter.java:230)
	at org.gradle.execution.taskgraph.DefaultTaskPlanExecutor$TaskExecutorWorker.processTask(DefaultTaskPlanExecutor.java:124)
	at org.gradle.execution.taskgraph.DefaultTaskPlanExecutor$TaskExecutorWorker.access$200(DefaultTaskPlanExecutor.java:80)
	at org.gradle.execution.taskgraph.DefaultTaskPlanExecutor$TaskExecutorWorker$1.execute(DefaultTaskPlanExecutor.java:105)
	at org.gradle.execution.taskgraph.DefaultTaskPlanExecutor$TaskExecutorWorker$1.execute(DefaultTaskPlanExecutor.java:99)
	at org.gradle.execution.taskgraph.DefaultTaskExecutionPlan.execute(DefaultTaskExecutionPlan.java:625)
	at org.gradle.execution.taskgraph.DefaultTaskExecutionPlan.executeWithTask(DefaultTaskExecutionPlan.java:580)
	at org.gradle.execution.taskgraph.DefaultTaskPlanExecutor$TaskExecutorWorker.run(DefaultTaskPlanExecutor.java:99)
	at org.gradle.internal.concurrent.ExecutorPolicy$CatchAndRecordFailures.onExecute(ExecutorPolicy.java:63)
	at org.gradle.internal.concurrent.ManagedExecutorImpl$1.run(ManagedExecutorImpl.java:46)
	at org.gradle.internal.concurrent.ThreadFactoryImpl$ManagedThreadRunnable.run(ThreadFactoryImpl.java:55)
Caused by: java.lang.RuntimeException: Job failed, see logs for details
	at com.android.build.gradle.internal.transforms.ProGuardTransform.transform(ProGuardTransform.java:196)
	at com.android.build.gradle.internal.pipeline.TransformTask$2.call(TransformTask.java:222)
	at com.android.build.gradle.internal.pipeline.TransformTask$2.call(TransformTask.java:218)
	at com.android.builder.profile.ThreadRecorder.record(ThreadRecorder.java:102)
	at com.android.build.gradle.internal.pipeline.TransformTask.transform(TransformTask.java:213)
	at org.gradle.internal.reflect.JavaMethod.invoke(JavaMethod.java:73)
	at org.gradle.api.internal.project.taskfactory.DefaultTaskClassInfoStore$IncrementalTaskAction.doExecute(DefaultTaskClassInfoStore.java:173)
	at org.gradle.api.internal.project.taskfactory.DefaultTaskClassInfoStore$StandardTaskAction.execute(DefaultTaskClassInfoStore.java:134)
	at org.gradle.api.internal.project.taskfactory.DefaultTaskClassInfoStore$StandardTaskAction.execute(DefaultTaskClassInfoStore.java:121)
	at org.gradle.api.internal.tasks.execution.ExecuteActionsTaskExecuter$1.run(ExecuteActionsTaskExecuter.java:122)
	at org.gradle.internal.progress.DefaultBuildOperationExecutor$RunnableBuildOperationWorker.execute(DefaultBuildOperationExecutor.java:336)
	at org.gradle.internal.progress.DefaultBuildOperationExecutor$RunnableBuildOperationWorker.execute(DefaultBuildOperationExecutor.java:328)
	at org.gradle.internal.progress.DefaultBuildOperationExecutor.execute(DefaultBuildOperationExecutor.java:197)
	at org.gradle.internal.progress.DefaultBuildOperationExecutor.run(DefaultBuildOperationExecutor.java:107)
	at org.gradle.api.internal.tasks.execution.ExecuteActionsTaskExecuter.executeAction(ExecuteActionsTaskExecuter.java:111)
	at org.gradle.api.internal.tasks.execution.ExecuteActionsTaskExecuter.executeActions(ExecuteActionsTaskExecuter.java:92)
	... 29 more
Caused by: java.io.IOException: Please correct the above warnings first.
	at proguard.Initializer.execute(Initializer.java:473)
	at proguard.ProGuard.initialize(ProGuard.java:233)
	at proguard.ProGuard.execute(ProGuard.java:98)
	at com.android.build.gradle.internal.transforms.BaseProguardAction.runProguard(BaseProguardAction.java:61)
	at com.android.build.gradle.internal.transforms.ProGuardTransform.doMinification(ProGuardTransform.java:253)
	at com.android.build.gradle.internal.transforms.ProGuardTransform.access$000(ProGuardTransform.java:63)
	at com.android.build.gradle.internal.transforms.ProGuardTransform$1.run(ProGuardTransform.java:173)
	at com.android.builder.tasks.Job.runTask(Job.java:47)
	at com.android.build.gradle.tasks.SimpleWorkQueue$EmptyThreadContext.runTask(SimpleWorkQueue.java:41)
	at com.android.builder.tasks.WorkQueue.run(WorkQueue.java:259)

```

No mini repro example. I will create one but it will take some time.

--- a/kategory-annotations-processor/src/main/java/kategory/derive/DerivingFileGenerator.kt
+++ b/kategory-annotations-processor/src/main/java/kategory/derive/DerivingFileGenerator.kt
@@ -136,7 +136,7 @@ class TypeclassInstanceGenerator(
             |}
             |
             |object ${instanceName}Implicits {
-            |  @JvmStatic fun instance(): $instanceName = ${receiverType}.Companion.${companionFactoryName}()
+            |  fun instance(): $instanceName = ${receiverType}.Companion.${companionFactoryName}()
             |}
             |
             |fun ${receiverType}.Companion.${companionFactoryName}(): $instanceName =

--- a/kategory-annotations-processor/src/main/java/kategory/instances/InstanceFileGenerator.kt
+++ b/kategory-annotations-processor/src/main/java/kategory/instances/InstanceFileGenerator.kt
@@ -174,7 +174,7 @@ class InstanceFileGenerator(
 
     private fun genImplicitObject(i: Instance): String = """
             |object ${i.implicitObjectName} {
-            |  @JvmStatic fun ${i.expandedTypeArgs()} instance(${i.expandedArgs}): ${i.name}${i.expandedTypeArgs()}${i.typeConstraints()} =
+            |  fun ${i.expandedTypeArgs()} instance(${i.expandedArgs}): ${i.name}${i.expandedTypeArgs()}${i.typeConstraints()} =
             |    object : ${i.name}${i.expandedTypeArgs()} {
             |${i.targetImplementations}
             |    }

--- a/kategory-annotations/src/main/java/kategory/Typeclass.kt
+++ b/kategory-annotations/src/main/java/kategory/Typeclass.kt
@@ -1,6 +1,8 @@
 package kategory
 
 import java.lang.IllegalArgumentException
+import java.lang.reflect.Method
+import java.lang.reflect.Modifier
 import java.lang.reflect.ParameterizedType
 import java.lang.reflect.Type
 import java.lang.reflect.TypeVariable
@@ -179,7 +181,7 @@ private fun instanceFromImplicitObject(t: InstanceParametrizedType): Any? {
     val allCompanionFunctions = globalInstanceProvider.methods
     val factoryFunction = allCompanionFunctions.find { it.name == "instance" }
     return if (factoryFunction != null) {
-        val values: List<Any> = factoryFunction.parameterTypes.mapIndexedNotNull { n, p ->
+        val values = factoryFunction.parameterTypes.mapIndexedNotNull { n, p ->
             if (Typeclass::class.java.isAssignableFrom(p)) {
                 val classifier = InstanceParametrizedType(p, p.typeParameters.toList())
                 val vType = reifyRawParameterizedType(t, classifier, n)
@@ -187,11 +189,29 @@ private fun instanceFromImplicitObject(t: InstanceParametrizedType): Any? {
                 if (value != null) registerInstance(vType, value)
                 value
             } else null
-        }
+        }.toTypedArray()
+
         factoryFunction.isAccessible = true
-        factoryFunction.invoke(null, *values.toTypedArray())
+
+        val instance =
+                // If the function has the modifier we don't need an instance to call it.
+                if (factoryFunction.isStatic()) null
+                // Otherwise we need to obtain the singleton object in order to call the
+                // 'instance' function on it
+                else globalInstanceProvider.fields.firstOrNull { it.name == "INSTANCE" }?.get(null)
+
+        if (values.isEmpty()) {
+            factoryFunction.invoke(instance)
+        } else {
+            factoryFunction.invoke(instance, *values)
+        }
+
     } else null
 }
+
+private fun Method.isStatic() =
+        Modifier.isStatic(modifiers)
+
 
 private fun reifyRawParameterizedType(carrier: InstanceParametrizedType, classifier: ParameterizedType, index: Int): InstanceParametrizedType =
         if (classifier.actualTypeArguments.any { it is TypeVariable<*> } && carrier.actualTypeArguments.size > index + 1) {
@@ -204,4 +224,4 @@ private fun reifyRawParameterizedType(carrier: InstanceParametrizedType, classif
         }
 
 private fun Type.asKotlinClass(): KClass<*>? =
-        if (this is Class<*>) this.kotlin else null
+        (this as? Class<*>)?.kotlin

--- a/kategory-annotations/src/main/java/kategory/Typeclass.kt
+++ b/kategory-annotations/src/main/java/kategory/Typeclass.kt
@@ -212,7 +212,6 @@ private fun instanceFromImplicitObject(t: InstanceParametrizedType): Any? {
 private fun Method.isStatic() =
         Modifier.isStatic(modifiers)
 
-
 private fun reifyRawParameterizedType(carrier: InstanceParametrizedType, classifier: ParameterizedType, index: Int): InstanceParametrizedType =
         if (classifier.actualTypeArguments.any { it is TypeVariable<*> } && carrier.actualTypeArguments.size > index + 1) {
             InstanceParametrizedType(classifier.rawType, listOf(carrier.actualTypeArguments[index + 1]))

--- a/kategory-core/src/main/kotlin/kategory/data/EitherT.kt
+++ b/kategory-core/src/main/kotlin/kategory/data/EitherT.kt
@@ -12,7 +12,7 @@ package kategory
 
         inline operator fun <reified F, A, B> invoke(value: HK<F, Either<A, B>>): EitherT<F, A, B> = EitherT(value)
 
-        @JvmStatic fun <F, A, B> pure(b: B, MF: Applicative<F>): EitherT<F, A, B> = right(b, MF)
+        fun <F, A, B> pure(b: B, MF: Applicative<F>): EitherT<F, A, B> = right(b, MF)
 
         fun <F, L, A, B> tailRecM(a: A, f: (A) -> EitherTKind<F, L, Either<A, B>>, MF: Monad<F>): EitherT<F, L, B> =
                 EitherT(MF.tailRecM(a, {
@@ -30,11 +30,11 @@ package kategory
                     }
                 }))
 
-        @JvmStatic fun <F, A, B> right(b: B, MF: Applicative<F>): EitherT<F, A, B> = EitherT(MF.pure(Right(b)))
+        fun <F, A, B> right(b: B, MF: Applicative<F>): EitherT<F, A, B> = EitherT(MF.pure(Right(b)))
 
-        @JvmStatic fun <F, A, B> left(a: A, MF: Applicative<F>): EitherT<F, A, B> = EitherT(MF.pure(Left(a)))
+        fun <F, A, B> left(a: A, MF: Applicative<F>): EitherT<F, A, B> = EitherT(MF.pure(Left(a)))
 
-        @JvmStatic inline fun <reified F, A, B> fromEither(value: Either<A, B>, MF: Applicative<F> = monad<F>()): EitherT<F, A, B> =
+        inline fun <reified F, A, B> fromEither(value: Either<A, B>, MF: Applicative<F> = monad<F>()): EitherT<F, A, B> =
                 EitherT(MF.pure(value))
 
         inline fun <reified F, L> functor(FF: Functor<F> = functor<F>()): Functor<EitherTKindPartial<F, L>> =

--- a/kategory-core/src/main/kotlin/kategory/data/Eval.kt
+++ b/kategory-core/src/main/kotlin/kategory/data/Eval.kt
@@ -48,21 +48,15 @@ sealed class Eval<out A> : EvalKind<A> {
 
         fun <A> pure(a: A): Eval<A> = Eval.now(a)
 
-
         fun <A> now(a: A) = Now(a)
-
 
         fun <A> later(f: () -> A) = Later(f)
 
-
         fun <A> always(f: () -> A) = Always(f)
-
 
         fun <A> defer(f: () -> Eval<A>): Eval<A> = Call(f)
 
-
         fun raise(t: Throwable): Eval<Nothing> = defer { throw t }
-
 
         val Unit: Eval<Unit> = Now(kotlin.Unit)
 

--- a/kategory-core/src/main/kotlin/kategory/data/Eval.kt
+++ b/kategory-core/src/main/kotlin/kategory/data/Eval.kt
@@ -48,30 +48,30 @@ sealed class Eval<out A> : EvalKind<A> {
 
         fun <A> pure(a: A): Eval<A> = Eval.now(a)
 
-        @JvmStatic
+
         fun <A> now(a: A) = Now(a)
 
-        @JvmStatic
+
         fun <A> later(f: () -> A) = Later(f)
 
-        @JvmStatic
+
         fun <A> always(f: () -> A) = Always(f)
 
-        @JvmStatic
+
         fun <A> defer(f: () -> Eval<A>): Eval<A> = Call(f)
 
-        @JvmStatic
+
         fun raise(t: Throwable): Eval<Nothing> = defer { throw t }
 
-        @JvmStatic
+
         val Unit: Eval<Unit> = Now(kotlin.Unit)
-        @JvmStatic
+
         val True: Eval<Boolean> = Now(true)
-        @JvmStatic
+
         val False: Eval<Boolean> = Now(false)
-        @JvmStatic
+
         val Zero: Eval<Int> = Now(0)
-        @JvmStatic
+
         val One: Eval<Int> = Now(1)
 
         fun <A, B> merge(

--- a/kategory-core/src/main/kotlin/kategory/data/Ior.kt
+++ b/kategory-core/src/main/kotlin/kategory/data/Ior.kt
@@ -73,7 +73,7 @@ typealias IorNel<A, B> = Ior<Nel<A>, B>
          * an [Ior.Left], [Ior.Right], or [Ior.Both] if [oa], [ob], or both are defined (respectively).
          */
 
-        @JvmStatic fun <A, B> fromOptions(oa: Option<A>, ob: Option<B>): Option<Ior<A, B>> = when (oa) {
+        fun <A, B> fromOptions(oa: Option<A>, ob: Option<B>): Option<Ior<A, B>> = when (oa) {
             is Some -> when (ob) {
                 is Some -> Some(Both(oa.value, ob.value))
                 is None -> Some(Left(oa.value))

--- a/kategory-core/src/main/kotlin/kategory/data/NonEmptyList.kt
+++ b/kategory-core/src/main/kotlin/kategory/data/NonEmptyList.kt
@@ -89,9 +89,9 @@ class NonEmptyList<out A> private constructor(
     override fun toString(): String = "NonEmptyList(all=$all)"
 
     companion object {
-        @JvmStatic fun <A> of(head: A, vararg t: A): NonEmptyList<A> = NonEmptyList(head, t.asList())
-        @JvmStatic fun <A> fromList(l: List<A>): Option<NonEmptyList<A>> = if (l.isEmpty()) None else Some(NonEmptyList(l))
-        @JvmStatic fun <A> fromListUnsafe(l: List<A>): NonEmptyList<A> = NonEmptyList(l)
+        fun <A> of(head: A, vararg t: A): NonEmptyList<A> = NonEmptyList(head, t.asList())
+        fun <A> fromList(l: List<A>): Option<NonEmptyList<A>> = if (l.isEmpty()) None else Some(NonEmptyList(l))
+        fun <A> fromListUnsafe(l: List<A>): NonEmptyList<A> = NonEmptyList(l)
 
         fun <A> pure(a: A): NonEmptyList<A> = a.nel()
 

--- a/kategory-core/src/main/kotlin/kategory/data/Option.kt
+++ b/kategory-core/src/main/kotlin/kategory/data/Option.kt
@@ -37,7 +37,6 @@ sealed class Option<out A> : OptionKind<A> {
             }
         }
 
-
         fun <A> fromNullable(a: A?): Option<A> = if (a != null) Some(a) else None
 
         operator fun <A> invoke(a: A): Option<A> = Some(a)

--- a/kategory-core/src/main/kotlin/kategory/data/Option.kt
+++ b/kategory-core/src/main/kotlin/kategory/data/Option.kt
@@ -37,7 +37,7 @@ sealed class Option<out A> : OptionKind<A> {
             }
         }
 
-        @JvmStatic
+
         fun <A> fromNullable(a: A?): Option<A> = if (a != null) Some(a) else None
 
         operator fun <A> invoke(a: A): Option<A> = Some(a)

--- a/kategory-core/src/main/kotlin/kategory/data/OptionT.kt
+++ b/kategory-core/src/main/kotlin/kategory/data/OptionT.kt
@@ -12,11 +12,11 @@ package kategory
 
         operator fun <F, A> invoke(value: HK<F, Option<A>>): OptionT<F, A> = OptionT(value)
 
-        @JvmStatic inline fun <reified F, A> pure(a: A, AF: Applicative<F> = kategory.applicative<F>()): OptionT<F, A> = OptionT(AF.pure(Some(a)))
+        inline fun <reified F, A> pure(a: A, AF: Applicative<F> = kategory.applicative<F>()): OptionT<F, A> = OptionT(AF.pure(Some(a)))
 
-        @JvmStatic inline fun <reified F> none(AF: Applicative<F> = kategory.applicative<F>()): OptionT<F, Nothing> = OptionT(AF.pure(None))
+        inline fun <reified F> none(AF: Applicative<F> = kategory.applicative<F>()): OptionT<F, Nothing> = OptionT(AF.pure(None))
 
-        @JvmStatic inline fun <reified F, A> fromOption(value: Option<A>, AF: Applicative<F> = kategory.applicative<F>()): OptionT<F, A> =
+        inline fun <reified F, A> fromOption(value: Option<A>, AF: Applicative<F> = kategory.applicative<F>()): OptionT<F, A> =
                 OptionT(AF.pure(value))
 
         fun <F, A, B> tailRecM(a: A, f: (A) -> OptionTKind<F, Either<A, B>>, MF: Monad<F>): OptionT<F, B> =

--- a/kategory-core/src/main/kotlin/kategory/data/Validated.kt
+++ b/kategory-core/src/main/kotlin/kategory/data/Validated.kt
@@ -11,25 +11,25 @@ typealias Invalid<E> = Validated.Invalid<E>
 
     companion object {
 
-        @JvmStatic fun <E, A> invalidNel(e: E): ValidatedNel<E, A> = Invalid(NonEmptyList(e, listOf()))
+        fun <E, A> invalidNel(e: E): ValidatedNel<E, A> = Invalid(NonEmptyList(e, listOf()))
 
-        @JvmStatic fun <E, A> validNel(a: A): ValidatedNel<E, A> = Validated.Valid(a)
+        fun <E, A> validNel(a: A): ValidatedNel<E, A> = Validated.Valid(a)
 
         /**
          * Converts a `Try<A>` to a `Validated<Throwable, A>`.
          */
-        @JvmStatic fun <A> fromTry(t: Try<A>): Validated<Throwable, A> = t.fold({ Invalid(it) }, { Valid(it) })
+        fun <A> fromTry(t: Try<A>): Validated<Throwable, A> = t.fold({ Invalid(it) }, { Valid(it) })
 
         /**
          * Converts an `Either<A, B>` to an `Validated<A, B>`.
          */
-        @JvmStatic fun <E, A> fromEither(e: Either<E, A>): Validated<E, A> = e.fold({ Invalid(it) }, { Valid(it) })
+        fun <E, A> fromEither(e: Either<E, A>): Validated<E, A> = e.fold({ Invalid(it) }, { Valid(it) })
 
         /**
          * Converts an `Option<B>` to an `Validated<A, B>`, where the provided `ifNone` values is returned on
          * the invalid of the `Validated` when the specified `Option` is `None`.
          */
-        @JvmStatic fun <E, A> fromOption(o: Option<A>, ifNone: () -> E): Validated<E, A> =
+        fun <E, A> fromOption(o: Option<A>, ifNone: () -> E): Validated<E, A> =
                 o.fold(
                         { Invalid(ifNone()) },
                         { Valid(it) }

--- a/kategory-core/src/main/kotlin/kategory/instances/CoproductInstances.kt
+++ b/kategory-core/src/main/kotlin/kategory/instances/CoproductInstances.kt
@@ -10,7 +10,7 @@ interface CoproductFunctorInstance<F, G> : Functor<CoproductKindPartial<F, G>> {
 }
 
 object CoproductFunctorInstanceImplicits {
-    @JvmStatic fun <F, G> instance(FF: Functor<F>, FG: Functor<G>): CoproductFunctorInstance<F, G> = object : CoproductFunctorInstance<F, G> {
+    fun <F, G> instance(FF: Functor<F>, FG: Functor<G>): CoproductFunctorInstance<F, G> = object : CoproductFunctorInstance<F, G> {
         override fun FF(): Functor<F> = FF
 
         override fun FG(): Functor<G> = FG
@@ -32,7 +32,7 @@ interface CoproductComonadInstance<F, G> : Comonad<CoproductKindPartial<F, G>> {
 }
 
 object CoproductComonadInstanceImplicits {
-    @JvmStatic fun <F, G> instance(CF: Comonad<F>, CG: Comonad<G>): CoproductComonadInstance<F, G> = object : CoproductComonadInstance<F, G> {
+    fun <F, G> instance(CF: Comonad<F>, CG: Comonad<G>): CoproductComonadInstance<F, G> = object : CoproductComonadInstance<F, G> {
         override fun CF(): Comonad<F> = CF
 
         override fun CG(): Comonad<G> = CG
@@ -52,7 +52,7 @@ interface CoproductFoldableInstance<F, G> : Foldable<CoproductKindPartial<F, G>>
 }
 
 object CoproductFoldableInstanceImplicits {
-    @JvmStatic fun <F, G> instance(FF: Foldable<F>, FG: Foldable<G>): CoproductFoldableInstance<F, G> = object : CoproductFoldableInstance<F, G> {
+    fun <F, G> instance(FF: Foldable<F>, FG: Foldable<G>): CoproductFoldableInstance<F, G> = object : CoproductFoldableInstance<F, G> {
         override fun FF(): Foldable<F> = FF
 
         override fun FG(): Foldable<G> = FG
@@ -75,7 +75,7 @@ interface CoproductTraverseInstance<F, G> : Traverse<CoproductKindPartial<F, G>>
 }
 
 object CoproductTraverseInstanceImplicits {
-    @JvmStatic fun <F, G> instance(TF: Traverse<F>, TG: Traverse<G>): CoproductTraverseInstance<F, G> = object : CoproductTraverseInstance<F, G> {
+    fun <F, G> instance(TF: Traverse<F>, TG: Traverse<G>): CoproductTraverseInstance<F, G> = object : CoproductTraverseInstance<F, G> {
         override fun TF(): Traverse<F> = TF
 
         override fun TG(): Traverse<G> = TG

--- a/kategory-core/src/main/kotlin/kategory/instances/EitherTInstances.kt
+++ b/kategory-core/src/main/kotlin/kategory/instances/EitherTInstances.kt
@@ -8,7 +8,6 @@ interface EitherTFunctorInstance<F, L> : Functor<EitherTKindPartial<F, L>> {
 }
 
 object EitherTFunctorInstanceImplicits {
-    @JvmStatic
     fun <F, L> instance(FF: Functor<F>): EitherTFunctorInstance<F, L> = object : EitherTFunctorInstance<F, L> {
         override fun FF(): Functor<F> = FF
     }
@@ -27,7 +26,6 @@ interface EitherTApplicativeInstance<F, L> : EitherTFunctorInstance<F, L>, Appli
 }
 
 object EitherTApplicativeInstanceImplicits {
-    @JvmStatic
     fun <F, L> instance(MF: Monad<F>): EitherTApplicativeInstance<F, L> = object : EitherTApplicativeInstance<F, L> {
         override fun FF(): Functor<F> = MF
 
@@ -41,14 +39,14 @@ interface EitherTMonadInstance<F, L> : EitherTApplicativeInstance<F, L>, Monad<E
             fa.ev().ap(ff,
                     MF())
 
-    override fun <A, B> flatMap(fa: EitherTKind<F, L, A>, f: (A) -> EitherTKind<F, L, B>): EitherT<F, L, B> = fa.ev().flatMap ({ f(it).ev() }, MF())
+    override fun <A, B> flatMap(fa: EitherTKind<F, L, A>, f: (A) -> EitherTKind<F, L, B>): EitherT<F, L, B> = fa.ev().flatMap({ f(it).ev() }, MF())
 
     override fun <A, B> tailRecM(a: A, f: (A) -> EitherTKind<F, L, Either<A, B>>): EitherT<F, L, B> =
             EitherT.tailRecM(a, f, MF())
 }
 
 object EitherTMonadInstanceImplicits {
-    @JvmStatic
+
     fun <F, L> instance(MF: Monad<F>): EitherTMonadInstance<F, L> = object : EitherTMonadInstance<F, L> {
         override fun FF(): Functor<F> = MF
 
@@ -70,7 +68,7 @@ interface EitherTMonadErrorInstance<F, L> : EitherTMonadInstance<F, L>, MonadErr
 }
 
 object EitherTMonadErrorInstanceImplicits {
-    @JvmStatic
+
     fun <F, L> instance(MF: Monad<F>): EitherTMonadErrorInstance<F, L> = object : EitherTMonadErrorInstance<F, L> {
         override fun FF(): Functor<F> = MF
 
@@ -88,7 +86,7 @@ interface EitherTFoldableInstance<F, L> : Foldable<EitherTKindPartial<F, L>> {
 }
 
 object EitherTFoldableInstanceImplicits {
-    @JvmStatic
+
     fun <F, L> instance(FF: Foldable<F>): EitherTFoldableInstance<F, L> = object : EitherTFoldableInstance<F, L> {
         override fun FFF(): Foldable<F> = FF
     }
@@ -105,7 +103,7 @@ interface EitherTTraverseInstance<F, L> : EitherTFunctorInstance<F, L>, EitherTF
 }
 
 object EitherTTraverseInstanceImplicits {
-    @JvmStatic
+
     fun <F, L> instance(TF: Traverse<F>): EitherTTraverseInstance<F, L> = object : EitherTTraverseInstance<F, L> {
         override fun FFF(): Foldable<F> = TF
 
@@ -123,7 +121,7 @@ interface EitherTSemigroupKInstance<F, L> : SemigroupK<EitherTKindPartial<F, L>>
 }
 
 object EitherTSemigroupKInstanceImplicits {
-    @JvmStatic
+
     fun <F, L> instance(MF: Monad<F>): EitherTSemigroupKInstance<F, L> = object : EitherTSemigroupKInstance<F, L> {
         override fun MF(): Monad<F> = MF
     }

--- a/kategory-core/src/main/kotlin/kategory/instances/NumberInstances.kt
+++ b/kategory-core/src/main/kotlin/kategory/instances/NumberInstances.kt
@@ -9,13 +9,13 @@ object ByteMonoid : Monoid<Byte>, Semigroup<Byte> by SGByte {
 }
 
 object ByteMonoidInstanceImplicits {
-    @JvmStatic fun instance(): ByteMonoid = ByteMonoid
+    fun instance(): ByteMonoid = ByteMonoid
 }
 
 private val SGByte: Semigroup<Byte> = NumberSemigroup({ one, two -> (one + two).toByte() })
 
 object ByteSemigroupInstanceImplicits {
-    @JvmStatic fun instance(): Semigroup<Byte> = SGByte
+    fun instance(): Semigroup<Byte> = SGByte
 }
 
 object ByteOrderInstance : Order<Byte> {
@@ -23,7 +23,7 @@ object ByteOrderInstance : Order<Byte> {
 }
 
 object ByteOrderInstanceImplicits {
-    @JvmStatic fun instance(): Order<Byte> = ByteOrderInstance
+    fun instance(): Order<Byte> = ByteOrderInstance
 }
 
 object ByteEqInstance: Eq<Byte> {
@@ -31,7 +31,7 @@ object ByteEqInstance: Eq<Byte> {
 }
 
 object ByteEqInstanceImplicits {
-    @JvmStatic fun instance(): ByteEqInstance = ByteEqInstance
+    fun instance(): ByteEqInstance = ByteEqInstance
 }
 
 object DoubleMonoid : Monoid<Double>, Semigroup<Double> by SGDouble {
@@ -39,13 +39,13 @@ object DoubleMonoid : Monoid<Double>, Semigroup<Double> by SGDouble {
 }
 
 object DoubleMonoidInstanceImplicits {
-    @JvmStatic fun instance(): DoubleMonoid = DoubleMonoid
+    fun instance(): DoubleMonoid = DoubleMonoid
 }
 
 private val SGDouble: Semigroup<Double> = NumberSemigroup(Double::plus)
 
 object DoubleSemigroupInstanceImplicits {
-    @JvmStatic fun instance(): Semigroup<Double> = SGDouble
+    fun instance(): Semigroup<Double> = SGDouble
 }
 
 object DoubleOrderInstance : Order<Double> {
@@ -53,7 +53,7 @@ object DoubleOrderInstance : Order<Double> {
 }
 
 object DoubleOrderInstanceImplicits {
-    @JvmStatic fun instance(): Order<Double> = DoubleOrderInstance
+    fun instance(): Order<Double> = DoubleOrderInstance
 }
 
 object DoubleEqInstance: Eq<Double> {
@@ -61,7 +61,7 @@ object DoubleEqInstance: Eq<Double> {
 }
 
 object DoubleEqInstanceImplicits {
-    @JvmStatic fun instance(): DoubleEqInstance = DoubleEqInstance
+    fun instance(): DoubleEqInstance = DoubleEqInstance
 }
 
 object IntMonoid : Monoid<Int>, Semigroup<Int> by SGInt {
@@ -69,13 +69,13 @@ object IntMonoid : Monoid<Int>, Semigroup<Int> by SGInt {
 }
 
 object IntMonoidInstanceImplicits {
-    @JvmStatic fun instance(): IntMonoid = IntMonoid
+    fun instance(): IntMonoid = IntMonoid
 }
 
 private val SGInt: Semigroup<Int> = NumberSemigroup(Int::plus)
 
 object IntSemigroupInstanceImplicits {
-    @JvmStatic fun instance(): Semigroup<Int> = SGInt
+    fun instance(): Semigroup<Int> = SGInt
 }
 
 object IntEqInstance: Eq<Int> {
@@ -83,7 +83,7 @@ object IntEqInstance: Eq<Int> {
 }
 
 object IntEqInstanceImplicits {
-    @JvmStatic fun instance(): IntEqInstance = IntEqInstance
+    fun instance(): IntEqInstance = IntEqInstance
 }
 
 object IntOrderInstance : Order<Int> {
@@ -91,7 +91,7 @@ object IntOrderInstance : Order<Int> {
 }
 
 object IntOrderInstanceImplicits {
-    @JvmStatic fun instance(): Order<Int> = IntOrderInstance
+    fun instance(): Order<Int> = IntOrderInstance
 }
 
 object LongMonoid : Monoid<Long>, Semigroup<Long> by SGLong {
@@ -99,13 +99,13 @@ object LongMonoid : Monoid<Long>, Semigroup<Long> by SGLong {
 }
 
 object LongMonoidInstanceImplicits {
-    @JvmStatic fun instance(): LongMonoid = LongMonoid
+    fun instance(): LongMonoid = LongMonoid
 }
 
 private val SGLong: Semigroup<Long> = NumberSemigroup(Long::plus)
 
 object LongSemigroupInstanceImplicits {
-    @JvmStatic fun instance(): Semigroup<Long> = SGLong
+    fun instance(): Semigroup<Long> = SGLong
 }
 
 object LongOrderInstance : Order<Long> {
@@ -113,7 +113,7 @@ object LongOrderInstance : Order<Long> {
 }
 
 object LongOrderInstanceImplicits {
-    @JvmStatic fun instance(): Order<Long> = LongOrderInstance
+    fun instance(): Order<Long> = LongOrderInstance
 }
 
 object LongEqInstance: Eq<Long> {
@@ -121,7 +121,7 @@ object LongEqInstance: Eq<Long> {
 }
 
 object LongEqInstanceImplicits {
-    @JvmStatic fun instance(): LongEqInstance = LongEqInstance
+    fun instance(): LongEqInstance = LongEqInstance
 }
 
 object ShortMonoid : Monoid<Short>, Semigroup<Short> by SGShort {
@@ -129,13 +129,13 @@ object ShortMonoid : Monoid<Short>, Semigroup<Short> by SGShort {
 }
 
 object ShortMonoidInstanceImplicits {
-    @JvmStatic fun instance(): ShortMonoid = ShortMonoid
+    fun instance(): ShortMonoid = ShortMonoid
 }
 
 private val SGShort: Semigroup<Short> = NumberSemigroup({ one, two -> (one + two).toShort() })
 
 object ShortSemigroupInstanceImplicits {
-    @JvmStatic fun instance(): Semigroup<Short> = SGShort
+    fun instance(): Semigroup<Short> = SGShort
 }
 
 object ShortOrderInstance : Order<Short> {
@@ -143,7 +143,7 @@ object ShortOrderInstance : Order<Short> {
 }
 
 object ShortOrderInstanceImplicits {
-    @JvmStatic fun instance(): Order<Short> = ShortOrderInstance
+    fun instance(): Order<Short> = ShortOrderInstance
 }
 
 object ShortEqInstance: Eq<Short> {
@@ -151,7 +151,7 @@ object ShortEqInstance: Eq<Short> {
 }
 
 object ShortEqInstanceImplicits {
-    @JvmStatic fun instance(): ShortEqInstance = ShortEqInstance
+    fun instance(): ShortEqInstance = ShortEqInstance
 }
 
 object FloatMonoid : Monoid<Float>, Semigroup<Float> by SGFloat {
@@ -159,13 +159,13 @@ object FloatMonoid : Monoid<Float>, Semigroup<Float> by SGFloat {
 }
 
 object FloatMonoidInstanceImplicits {
-    @JvmStatic fun instance(): FloatMonoid = FloatMonoid
+    fun instance(): FloatMonoid = FloatMonoid
 }
 
 private val SGFloat: Semigroup<Float> = NumberSemigroup(Float::plus)
 
 object FloatSemigroupInstanceImplicits {
-    @JvmStatic fun instance(): Semigroup<Float> = SGFloat
+    fun instance(): Semigroup<Float> = SGFloat
 }
 
 object FloatOrderInstance : Order<Float> {
@@ -173,7 +173,7 @@ object FloatOrderInstance : Order<Float> {
 }
 
 object FloatOrderInstanceImplicits {
-    @JvmStatic fun instance(): Order<Float> = FloatOrderInstance
+    fun instance(): Order<Float> = FloatOrderInstance
 }
 
 object FloatEqInstance: Eq<Float> {
@@ -181,5 +181,5 @@ object FloatEqInstance: Eq<Float> {
 }
 
 object FloatEqInstanceImplicits {
-    @JvmStatic fun instance(): FloatEqInstance = FloatEqInstance
+    fun instance(): FloatEqInstance = FloatEqInstance
 }

--- a/kategory-core/src/main/kotlin/kategory/instances/NumberInstancesJava.kt
+++ b/kategory-core/src/main/kotlin/kategory/instances/NumberInstancesJava.kt
@@ -3,97 +3,97 @@ package java_lang
 import kategory.*
 
 object IntegerMonoidInstanceImplicits {
-    @JvmStatic fun instance(): IntMonoid = IntMonoid
+    fun instance(): IntMonoid = IntMonoid
 }
 
 object IntegerSemigroupInstanceImplicits {
-    @JvmStatic fun instance(): Semigroup<Int> = IntMonoid
+    fun instance(): Semigroup<Int> = IntMonoid
 }
 
 object IntegerEqInstanceImplicits {
-    @JvmStatic fun instance(): IntEqInstance = IntEqInstance
+    fun instance(): IntEqInstance = IntEqInstance
 }
 
 object IntegerOrderInstanceImplicits {
-    @JvmStatic fun instance(): IntOrderInstance = IntOrderInstance
+    fun instance(): IntOrderInstance = IntOrderInstance
 }
 
 object LongMonoidInstanceImplicits {
-    @JvmStatic fun instance(): LongMonoid = LongMonoid
+    fun instance(): LongMonoid = LongMonoid
 }
 
 object LongSemigroupInstanceImplicits {
-    @JvmStatic fun instance(): Semigroup<Long> = LongMonoid
+    fun instance(): Semigroup<Long> = LongMonoid
 }
 
 object LongOrderInstanceImplicits {
-    @JvmStatic fun instance(): LongOrderInstance = LongOrderInstance
+    fun instance(): LongOrderInstance = LongOrderInstance
 }
 
 object LongEqInstanceImplicits {
-    @JvmStatic fun instance(): LongEqInstance = LongEqInstance
+    fun instance(): LongEqInstance = LongEqInstance
 }
 
 object ShortMonoidInstanceImplicits {
-    @JvmStatic fun instance(): ShortMonoid = ShortMonoid
+    fun instance(): ShortMonoid = ShortMonoid
 }
 
 object ShortSemigroupInstanceImplicits {
-    @JvmStatic fun instance(): Semigroup<Short> = ShortMonoid
+    fun instance(): Semigroup<Short> = ShortMonoid
 }
 
 object ShortOrderInstanceImplicits {
-    @JvmStatic fun instance(): ShortOrderInstance = ShortOrderInstance
+    fun instance(): ShortOrderInstance = ShortOrderInstance
 }
 
 object ShortEqInstanceImplicits {
-    @JvmStatic fun instance(): ShortEqInstance = ShortEqInstance
+    fun instance(): ShortEqInstance = ShortEqInstance
 }
 
 object ByteMonoidInstanceImplicits {
-    @JvmStatic fun instance(): ByteMonoid = ByteMonoid
+    fun instance(): ByteMonoid = ByteMonoid
 }
 
 object ByteSemigroupInstanceImplicits {
-    @JvmStatic fun instance(): Semigroup<Byte> = ByteMonoid
+    fun instance(): Semigroup<Byte> = ByteMonoid
 }
 
 object ByteOrderInstanceImplicits {
-    @JvmStatic fun instance(): ByteOrderInstance = ByteOrderInstance
+    fun instance(): ByteOrderInstance = ByteOrderInstance
 }
 
 object ByteEqInstanceImplicits {
-    @JvmStatic fun instance(): ByteEqInstance = ByteEqInstance
+    fun instance(): ByteEqInstance = ByteEqInstance
 }
 
 object DoubleMonoidInstanceImplicits {
-    @JvmStatic fun instance(): DoubleMonoid = DoubleMonoid
+    fun instance(): DoubleMonoid = DoubleMonoid
 }
 
 object DoubleSemigroupInstanceImplicits {
-    @JvmStatic fun instance(): Semigroup<Double> = DoubleMonoid
+    fun instance(): Semigroup<Double> = DoubleMonoid
 }
 
 object DoubleOrderInstanceImplicits {
-    @JvmStatic fun instance(): DoubleOrderInstance = DoubleOrderInstance
+    fun instance(): DoubleOrderInstance = DoubleOrderInstance
 }
 
 object DoubleEqInstanceImplicits {
-    @JvmStatic fun instance(): DoubleEqInstance = DoubleEqInstance
+    fun instance(): DoubleEqInstance = DoubleEqInstance
 }
 
 object FloatMonoidInstanceImplicits {
-    @JvmStatic fun instance(): FloatMonoid = FloatMonoid
+    fun instance(): FloatMonoid = FloatMonoid
 }
 
 object FloatSemigroupInstanceImplicits {
-    @JvmStatic fun instance(): Semigroup<Float> = FloatMonoid
+    fun instance(): Semigroup<Float> = FloatMonoid
 }
 
 object FloatOrderInstanceImplicits {
-    @JvmStatic fun instance(): FloatOrderInstance = FloatOrderInstance
+    fun instance(): FloatOrderInstance = FloatOrderInstance
 }
 
 object FloatEqInstanceImplicits {
-    @JvmStatic fun instance(): FloatEqInstance = FloatEqInstance
+    fun instance(): FloatEqInstance = FloatEqInstance
 }

--- a/kategory-core/src/main/kotlin/kategory/instances/StringInstances.kt
+++ b/kategory-core/src/main/kotlin/kategory/instances/StringInstances.kt
@@ -5,7 +5,7 @@ object StringSemigroupInstance : Semigroup<String> {
 }
 
 object StringSemigroupInstanceImplicits {
-    @JvmStatic
+
     fun instance(): StringSemigroupInstance = StringSemigroupInstance
 }
 
@@ -16,7 +16,7 @@ object StringMonoidInstance : Monoid<String> {
 }
 
 object StringMonoidInstanceImplicits {
-    @JvmStatic
+
     fun instance(): StringMonoidInstance = StringMonoidInstance
 }
 
@@ -25,6 +25,6 @@ object StringEqInstance : Eq<String> {
 }
 
 object StringEqInstanceImplicits {
-    @JvmStatic
+
     fun instance(): StringEqInstance = StringEqInstance
 }

--- a/kategory-core/src/main/kotlin/kategory/instances/StringInstancesJava.kt
+++ b/kategory-core/src/main/kotlin/kategory/instances/StringInstancesJava.kt
@@ -3,14 +3,14 @@ package java_lang
 import kategory.*
 
 object StringMonoidInstanceImplicits {
-    @JvmStatic fun instance(): StringMonoidInstance = StringMonoidInstance
+    fun instance(): StringMonoidInstance = StringMonoidInstance
 }
 
 object StringSemigroupInstanceImplicits {
-    @JvmStatic fun instance(): StringSemigroupInstance = StringSemigroupInstance
+    fun instance(): StringSemigroupInstance = StringSemigroupInstance
 }
 
 object StringEqInstanceImplicits {
-    @JvmStatic
+
     fun instance(): StringEqInstance = StringEqInstance
 }

--- a/kategory-core/src/main/kotlin/kategory/instances/UnitInstances.kt
+++ b/kategory-core/src/main/kotlin/kategory/instances/UnitInstances.kt
@@ -9,7 +9,7 @@ object UnitSemigroupInstance: Semigroup<Unit> {
 }
 
 object UnitSemigroupInstanceImplicits {
-    @JvmStatic fun instance(): Semigroup<Unit> = UnitSemigroupInstance
+    fun instance(): Semigroup<Unit> = UnitSemigroupInstance
 }
 
 object UnitMonoidInstance: Monoid<Unit>, Semigroup<Unit> by UnitSemigroupInstance {
@@ -17,7 +17,7 @@ object UnitMonoidInstance: Monoid<Unit>, Semigroup<Unit> by UnitSemigroupInstanc
 }
 
 object UnitMonoidInstanceImplicits {
-    @JvmStatic fun instance(): Monoid<Unit> = UnitMonoidInstance
+    fun instance(): Monoid<Unit> = UnitMonoidInstance
 }
 
 object UnitEqInstance : Eq<Unit> {
@@ -25,5 +25,5 @@ object UnitEqInstance : Eq<Unit> {
 }
 
 object UnitEqInstanceImplicits {
-    @JvmStatic fun instance(): Eq<Unit> = UnitEqInstance
+    fun instance(): Eq<Unit> = UnitEqInstance
 }

--- a/kategory-core/src/test/kotlin/kategory/data/SortedMapKWTest.kt
+++ b/kategory-core/src/test/kotlin/kategory/data/SortedMapKWTest.kt
@@ -2,7 +2,6 @@ package kategory
 
 import io.kotlintest.KTestJUnitRunner
 import io.kotlintest.matchers.shouldNotBe
-import io.kotlintest.properties.forAll
 import org.junit.runner.RunWith
 
 @RunWith(KTestJUnitRunner::class)
@@ -10,7 +9,7 @@ class SortedMapKWTest : UnitSpec() {
 
     val EQ: Eq<HK2<SortedMapKWHK, String, Int>> = object : Eq<HK2<SortedMapKWHK, String, Int>> {
         override fun eqv(a: HK2<SortedMapKWHK, String, Int>, b: HK2<SortedMapKWHK, String, Int>): Boolean =
-            a.ev().get("key") == b.ev().get("key")
+            a.ev()["key"] == b.ev()["key"]
     }
 
     init {

--- a/kategory-effects/src/main/kotlin/kategory/effects/data/Duration.kt
+++ b/kategory-effects/src/main/kotlin/kategory/effects/data/Duration.kt
@@ -7,7 +7,7 @@ data class Duration(val amount: Long, val timeUnit: TimeUnit) {
 
     companion object {
         // Actually limited to 9223372036854775807 days, so unless you are very patient, it is unlimited ;-)
-        @JvmStatic val INFINITE = Duration(amount = Long.MAX_VALUE, timeUnit = TimeUnit.DAYS)
+        val INFINITE = Duration(amount = Long.MAX_VALUE, timeUnit = TimeUnit.DAYS)
     }
 }
 

